### PR TITLE
[espnet3-5] Add parallel module and collect_stats

### PR DESCRIPTION
**Pull request by Masao-Someki**
## What did you change?

- Added a `espnet3.parallel` module based on `Dask` for supporting parallel processing on various environments.
- Added a dask-based `collect_stats` module.
- The `collect_stats` function supports multiple execution modes: local, parallel (Dask-based), and multiple iterator (sharded datasets).

------

## Why did you make this change?

- To enable robust and flexible collection of feature-level statistics required for normalization or data analysis.
- `espnet3/parallel` is also required for parallel decoding in evaluation stage.

------

## Is your PR small enough?

Yes.

This PR mainly adds 2 files (`espnet3/parallel/parallel.py` and `espnet3/collect_stats.py`) with around 700 lines

------

## Additional Context

* #6133
* #6139 
* #6167 
* #6171 
* #6172 
* #6173

Note: Since I couldn't find a way to run CI with parallel processing, this PR does not contain unit tests.

### CODE REVIEW

**Comment by Masao-Someki** [comment]
The last 3 commits are the item of this PR. Other commits should be merged into `espnet3` branch by the other PRs.

**Comment by Fhrozen** [comment]
This pull request introduces support for ESPnet-3 across various CI workflows and adds new functionality in the ESPnet-3 codebase. Key changes include extending CI workflows to handle ESPnet-3, introducing new dataset handling capabilities, and updating type annotations for compatibility with Hydra configurations.

### CI Workflow Updates for ESPnet-3:
* [`.github/workflows/ci_on_debian11.yml`](diffhunk://#diff-b13dff138f65e577647f126f573d36d77fe805c8018d819d439da401f913456cR7-R14): Added `espnet3` branch support and updated job names and scripts to include ESPnet-3 testing. [[1]](diffhunk://#diff-b13dff138f65e577647f126f573d36d77fe805c8018d819d439da401f913456cR7-R14) [[2]](diffhunk://#diff-b13dff138f65e577647f126f573d36d77fe805c8018d819d439da401f913456cR47-R52)
* `.github/workflows/ci_on_macos.yml`, `.github/workflows/ci_on_ubuntu.yml`, `.github/workflows/ci_on_windows.yml`: Extended CI workflows to include `espnet3` branch handling. [[1]](diffhunk://#diff-e9868fff133d797c4c6a9e2ad3cbc3fc55bfd09b1a2cc1324218b5609567475fR7-R11) [[2]](diffhunk://#diff-48d19e7512e7e9ec72c5503d3178348ad59275991f8b8334269329766e343de9R7-R11) [[3]](diffhunk://#diff-07e2a7a53f167a4bfe47fd72cc304f5071279a86c297985caaabd11ff35da7b1R7-R11)
* [`.github/workflows/ci_on_ubuntu.yml`](diffhunk://#diff-48d19e7512e7e9ec72c5503d3178348ad59275991f8b8334269329766e343de9L23-R25): Enhanced matrix configurations for Python and PyTorch versions, added PR label checks for ESPnet-3, and adjusted testing scripts. [[1]](diffhunk://#diff-48d19e7512e7e9ec72c5503d3178348ad59275991f8b8334269329766e343de9L23-R25) [[2]](diffhunk://#diff-48d19e7512e7e9ec72c5503d3178348ad59275991f8b8334269329766e343de9L178-R181) [[3]](diffhunk://#diff-48d19e7512e7e9ec72c5503d3178348ad59275991f8b8334269329766e343de9R198-R200)

### ESPnet-3 Codebase Enhancements:
* `espnet3/__init__.py`, `espnet3/data/__init__.py`: Introduced new modules for ESPnet-3, including `LitESPnetModel`, `ESPnet3LightningTrainer`, and dataset-related utilities. [[1]](diffhunk://#diff-643d3f639d07285307d53dd2149650dbd2273b36a4d76d6fe4040c30d8247b69R1-R2) [[2]](diffhunk://#diff-95851550ab39f5d69422356441d617837e9d578e242941a71f3d38a096c17bedR1-R2)
* [`espnet3/data/data_organizer.py`](diffhunk://#diff-294170c415c4d9ca737ae8e58c78c357365c6b5e24418a4fb925889895fdc229R1-R197): Added `DataOrganizer` class to unify training, validation, and test dataset handling, along with `DatasetConfig` for flexible configuration.

### Type Annotation Updates:
* [`espnet2/samplers/build_batch_sampler.py`](diffhunk://#diff-8d83b88995d9bd7306fcf8e5d9d6c8a4705c36d461d9b1700300a15dfaa7c753R3): Updated `shape_files` parameter to support `ListConfig` type from Hydra for compatibility with advanced configurations. [[1]](diffhunk://#diff-8d83b88995d9bd7306fcf8e5d9d6c8a4705c36d461d9b1700300a15dfaa7c753R3) [[2]](diffhunk://#diff-8d83b88995d9bd7306fcf8e5d9d6c8a4705c36d461d9b1700300a15dfaa7c753L77-R78)

### Testing Script Updates:
* [`ci/test_python_espnet3.sh`](diffhunk://#diff-c35bd8cb103cc461d2a223e9296562eb72721f83f578c145eb338349b326d6d4L13-R16): Renamed from `test_python_espnetez.sh` and updated to test ESPnet-3-specific components. Adjusted `pycodestyle` exclusions and replaced test paths.

**Comment by Fhrozen** [comment]
@dosu-bot

**Comment by sw005320** [comment]
@Masao-Someki, please work on this PR

**Comment by sw005320** [comment]
@Emrys365, can you review this PR?

File: espnet3/collect_stats.py
Comment (sw005320) [inline espnet3/collect_stats.py:1]: This function tightly integrates the Dask-based multi-processing and statistics collection
Can we refactor to separate Dask-based multi-processing and (single-thread) statistics collection, and call multiple (single-thread) statistics computations from Dask-based multi-processing?

The reason is that when people touch statistics computation (or other processing), they also have to understand the Dask-based multi-processing.

Comment (Masao-Someki) [reply espnet3/collect_stats.py:1]: For local parallelism, `torch.distributed` and Dask both work, but in cluster environments like slurm, Dask can automatically submit and scale jobs based on available resources. `torch.distributed` requires us to pre-request all resources for parallelizing jobs before running.

> Use PyTorch distributed for parallelizing deep learning training.
> Use Dask for parallelizing data processing and any workload not centered on neural network training.

Yes, this is what I'm thinking of.

> Can we use Dask for inference, too?

Yes! Since we can request GPU jobs, we can run inference with it.

>Can we refactor to separate Dask-based multi-processing and (single-thread) statistics collection, and call multiple (single-thread) statistics computations from Dask-based multi-processing?
>
> The reason is that when people touch statistics computation (or other processing), they also have to understand the Dask-based multi-processing.

This makes sense, I agree with this approach and will proceed with the refactoring.

**Comment by sw005320** [comment]
There is a CI issue

**Comment by Fhrozen** [comment]
I think that there are many changes that should be handled in a different pr. It will be better if you first update the espnet 3 branch from the master, before any additional change.

**Comment by sw005320** [comment]
This is my suggestion.
espnet3's test seems to be an issue, which can be solved by merging the main repo, but if it is too much, we can revert it

**Comment by sw005320** [comment]
I'd recommend you make a separate PR to merge master to espnet3
This PR should be independent of this

File: espnet3/collect_stats.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    Entry point to collect feature statistics from a dataset.
+
+    - Local single-process path
+    - Dask parallel path
+    - Multi-iterator (sharded) path (parallel only; no raw feats saved)
+    """
+    mode_config = getattr(dataloader_config, mode)
+    if getattr(mode_config, "multiple_iterator", False):
+        if parallel_config is None:
+            raise RuntimeError("You should set parallel config with multiple iterator.")
```
Comment (Emrys365) [inline espnet3/collect_stats.py:37]: Is there any documentation on `parallel_config` that can be referred to here? I think it is needed for people to learn how to set this config.

File: espnet3/collect_stats.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    model_config,
+    dataset_config,
+    dataloader_config,
+    mode: str,
+    output_dir: Path,
+    task: Optional[str] = None,
+    parallel_config: Optional[dict] = None,
+    write_collected_feats: bool = True,
+    batch_size: int = 4,
+):
```
Comment (Emrys365) [inline espnet3/collect_stats.py:26]: Can you add explanations of these arguments in the docstring? It is not clear to me what `mode` and `parallel_config` mean and how they should be set.

File: espnet3/parallel/parallel.py [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+
+    try:
+        _check_conflict_client_side(func, kwargs, client)
+        wrapped_func = wrap_func_with_worker_env(func)
+        futures = client.map(wrapped_func, args, **kwargs)
+        for future in as_completed(futures):
+            yield future.result()
+    finally:
+        if internal:
+            client_cm.__exit__(None, None, None)
```
Comment (Emrys365) [inline espnet3/parallel/parallel.py:532]: The `parallel_map` and `parallel_for` functions only differ in how they return the result. Could you simplify the code by merging the common part?

File: espnet3/utils/collect_stats_local.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        k: (v.detach().cpu().numpy() if hasattr(v, "detach") else v)
+        for k, v in feats.items()
+    }
+
+    stats = defaultdict(lambda: {"sum": 0, "sq": 0, "count": 0})
+    shape_info = defaultdict(dict)
+
+    uid_list = list(uids)
+    for b_idx, uid in enumerate(uid_list):
+        for feat_key in list(feats.keys()):
```
Comment (Emrys365) [inline espnet3/utils/collect_stats_local.py:56]:
```suggestion
        for feat_key in feats.keys():
```

File: espnet3/utils/collect_stats_local.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+):
+    """
+    Process a batch of dataset indices to compute feature statistics.
+
+    Returns:
+        If write_collected_feats:
+            Tuple[stats, shape_info, feats]
+        else:
+            Tuple[stats, shape_info]
+    """
```
Comment (Emrys365) [inline espnet3/utils/collect_stats_local.py:34]: The function should respect the `extract_feats_in_collect_stats` attribute defined in `model`. If the attribute is `False`, feature-related stat calculations should be skipped, as done in [espnet2/tasks/abs_task.py#L1405-L1411](https://github.com/espnet/espnet/blob/master/espnet2/tasks/abs_task.py#L1405-L1411) and [espnet2/main_funcs/collect_stats.py#L66](https://github.com/espnet/espnet/blob/master/espnet2/main_funcs/collect_stats.py#L66)

**Comment by sw005320** [comment]
@Masao-Someki, can you work on it?

**Comment by sw005320** [comment]
@Masao-Someki, please restart the PRs by merging your updates and fixing the conflicts.

File: espnet3/collect_stats.py (hunk 1 above)
Comment (Masao-Someki) [reply espnet3/collect_stats.py:37]: Thank you. Since the documentation isn’t on PR yet, I’ll add a warning comment that says “see docs for more details.”

**Comment by Masao-Someki** [comment]
Since this PR has become complicated, I’d like to restart from the current `espnet3` branch.
Merging `espnet3` into this branch is not straightforward, because we previously merged `master` and then reverted it, which left that history in place. As a result, for example, the Gemini review config file should appear after merging `espnet3`, but the change is not reflected in this branch, because the revert commit actually deleted the config file.

I will close this PR and create a new one from the current espnet3 branch.

File: espnet3/utils/collect_stats_local.py (hunk 5 above)
Comment (Masao-Someki) [reply espnet3/utils/collect_stats_local.py:34]: Thank you, since the collect stats process is optional in espnet-3, I think we can leave this as-is. Users who don’t want to run it can simply skip calling the `collect_stats` function.
